### ISSUE TITLE
rule: require `setup` on `<script>`

### ISSUE THREAD

**Issue by privatenumber** [vuejs/eslint-plugin-vue#2328]
**Please describe what the rule should do:**

The rule should enforce that if there's a `<script>`, the `setup` attribute should be on it.

**What category should the rule belong to?**

- [x] Enforces code style (layout)
- [ ] Warns about a potential error (problem)
- [ ] Suggests an alternate way of doing something (suggestion)
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**

Should also warn against non-`setup` `<script>` in a codebase that uses `setup`:
```vue
<script>
export default {
   mounted() {...}
}
</script>
```

I also saw a case where a dev was confused why they couldn't use `<Component>`, and it was because they forgot to add `setup` to `<script>`:
```vue
<script>
import Component from './component.vue';
</script>
```
(In this case, it might also make sense to propose a separate rule that checks for `<script>` tags with no default exports)

**Additional context**

For implementation, I was wondering if it might make sense to update [`block-lang`](https://eslint.vuejs.org/rules/block-lang.html) to something like `block-attrs` where both `lang` & `setup` could be specified as required attributes.

**Comment by ota-meshi** [comment]
Thank you for rule suggestion!
However, I think you might be able to use the `vue/component-api-style` rule.
https://eslint.vuejs.org/rules/component-api-style.html

[DEMO](https://ota-meshi.github.io/eslint-plugin-vue-demo/#eJx1lMFu2zAMhl/F8GkrKgTYMSj2InMPtEzbamXRI6UkXtB3H2U3dpx1QABHHyWK/EnqWlpqsDyWL2LZjbH4WYWiODwVKN6FWFTlKeHB0jBSwBANjM5InDxW5bH4VZXITFyVz/n/4sEIxjRW5etr8XSoAl5G4lg02ELysbhm98VAKURsvn0vrh9V0N/LYTmst5fPJSePUh6vt7uHfHPjGG10Jw32x/NseZOLSYJiTsAyU0tBIicbiY2kEXmmra4+T1OYSYcxIhvWSHkhgQzIFKwZmQYnaPCCNqmfm9WCwgatB4bsR1aDBxE9LK5bXWW9gNEE7MwfZNp4aP7ZqiF/xUBzzvvdGvViiUxeI+/wcoMN1qnrkLe1x4hZlZWoFga4kx1YIh9wqJH3FvSarGt37B2n+03eWdBLsiw3isMYp93C2F7lslnr+bK9cYRchTU3vDzIgJfIYGoijxDyTfFmacH72DOlrl9R0uLtz3eeavAP0A25Hx9hCBriV9V14QTeNYvg40pZ10k3m3PvIsoIdlXBk4pKrTYSWid3tdO20kQaF7r/yRLwbGQaavIroUA2ahIamxv0i2JhXO+i+k118X71MG++LbSVI8VJa1cn56PbsmJccsUNaHpmzmPdJOjbB53kq7mRHho6K9TRc5pUYwIMd250EiR3H8PWQLF3YmrUycS7QVVDCvpSbAvVHGeXgz4eTl8k3GyMoDrW/g4JtGhaF1SS6YHqKML9JH1iGjNUYdWVC1qbza5PS2M81LjJ+wnX9yYjlQm14jXYd8YWGcPWCzejjortH+G+lGcXlx2Mv5M+VmZy6JuZpDyMEmCJfenHXFbKQn18/AXYPPh1)

However, note that depending on the definition of `export default` this rule will not report it. This is because some are allowed when used together with `<script setup>` and `<script>`. It is considered an abbreviation for `<script setup>`.

```vue
<script >
  /* eslint "vue/component-api-style": ["error", ["script-setup"]] */
export default {
  name: 'foo'
}
</script>
```

**Comment by privatenumber** [comment]
My mistake; I searched variations of "script setup" on the docs but didn't think to search for the Composition API.

Thanks for the careful explanation @ota-meshi !

Is there a rule that catches `<script>` tags with no `export default`? I know it's technically allowed, but I think a warning could help catch mistakes during development.

**Comment by ota-meshi** [comment]
> Is there a rule that catches <script> tags with no export default? I know it's technically allowed, but I think a warning could help catch mistakes during development.

I don't think this plugin has such rules right now.

I think it might be good to have a rule that prohibits `<script>` blocks that don't have `export`, and a rule that prefers `defineOptions` 🤔
